pkg/cache/v3/linear.go

```diff
@@ -241,82 +267,114 @@ func (cache *LinearCache) computeResourceChange(sub Subscription, ignoreReturned
+	// TODO(valerian-roche): review if using https://pkg.go.dev/github.com/mitchellh/hashstructure/v2 is possible to remove the need to sort.
+	keys := make([]string, 0, len(versionMap))
+	for key := range versionMap {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	mapHasher := fnv.New64()
+
+	buffer := make([]byte, 0, 8)
+	itemHasher := fnv.New64()
+	for _, key := range keys {
+		buffer = buffer[:0]
+		itemHasher.Reset()
+		itemHasher.Write([]byte(key))
+		mapHasher.Write(itemHasher.Sum(buffer))
+		buffer = buffer[:0]
+		itemHasher.Reset()
+		itemHasher.Write([]byte(versionMap[key]))
+		mapHasher.Write(itemHasher.Sum(buffer))
+	}
+	buffer = buffer[:0]
+	return hex.EncodeToString(mapHasher.Sum(buffer))
+}
+
 func (cache *LinearCache) computeSotwResponse(watch ResponseWatch, ignoreReturnedResources bool) (*RawResponse, error) {
-	changedResources, removedResources, err := cache.computeResourceChange(watch.subscription, ignoreReturnedResources, false)
+	changedResources, removedResources, err := cache.computeResourceChange(watch.subscription, ignoreReturnedResources, cache.useStableVersionsInSotw)
 	if err != nil {
 		return nil, err
 	}
-
 	if len(changedResources) == 0 && len(removedResources) == 0 && !ignoreReturnedResources {
 		// Nothing changed.
 		return nil, nil
 	}
 
-	returnedVersions := make(map[string]string, len(watch.subscription.ReturnedResources()))
-	// Clone the current returned versions. The cache should not alter the subscription
-	for resourceName, version := range watch.subscription.ReturnedResources() {
-		returnedVersions[resourceName] = version
-	}
-
-	cacheVersion := cache.getVersion()
-	var resources []types.ResourceWithTTL
+	// In sotw the list of resources to actually return depends on:
+	//  - whether the type requires full-state in each reply (e.g. lds, cds).
+	//  - whether the request is wildcard.
+	// resourcesToReturn will include all the resource names to reply based on the changes detected.
+	var resourcesToReturn []string
 
 	switch {
 	// Depending on the type, the response will only include changed resources or all of them
 	case !ResourceRequiresFullStateInSotw(cache.typeURL):
-		// changedResources is already filtered based on the subscription.
-		// TODO(valerian-roche): if the only change is a removal in the subscription,
-		// or a watched resource getting deleted, this might send an empty reply.
-		// While this does not violate the protocol, we might want to avoid it.
-		resources = make([]types.ResourceWithTTL, 0, len(changedResources))
-		for _, resourceName := range changedResources {
-			cachedResource := cache.resources[resourceName]
-			resources = append(resources, types.ResourceWithTTL{Resource: cachedResource.Resource})
-			returnedVersions[resourceName] = cachedResource.cacheVersion
+		if !ignoreReturnedResources && len(changedResources) == 0 {
+			// If the request is not the initial one, and the type does not require full updates,
+			// do not return if nothing is to be set.
```

Comment:
I'm a bit confused by this message. 
`Do not return` I think implies that we are not sending a response here, but then the comment `For full state resource an empty response has a semantic meaning`, implies that we are sending an empty response here.

Reply:
This comment is within `!ResourceRequiresFullStateInSotw` and does not apply to full state resources